Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,679 @@
+    ref->ack_num = htonl(tcpdata.sequence_number) + 1;
+    ref->ack_base = ref->ack_num;
+    ref->seq_num = 0x1000000;
+    ref->window_size = htons(tcpdata.window_size);
+    ref->type = IPPROTO_TCP;
+    for (auto& tcp_buf : ref->tcp_buffers)
+      tcp_buf.used = false;
+    ref->from.sin_addr.s_addr = Common::BitCast<u32>(ipdata.destination_addr);
+    ref->from.sin_port = tcpdata.destination_port;
+    ref->to.sin_addr.s_addr = Common::BitCast<u32>(ipdata.source_addr);
+    ref->to.sin_port = tcpdata.source_port;
+    ref->bba_mac = Common::BitCastPtr<Common::MACAddress>(&m_eth_ref->mBbaMem[BBA_NAFR_PAR0]);
+    ref->my_mac = m_fake_mac;
+    ref->tcp_socket.setBlocking(false);
+
+    // reply with a sin_ack
+    Common::TCPPacket result(ref->bba_mac, ref->my_mac, ref->from, ref->to, ref->seq_num,
+                             ref->ack_num, TCP_FLAG_SIN | TCP_FLAG_ACK);
+
+    result.tcp_options = std::vector<u8>{0x02, 0x04, 0x05, 0xb4, 0x01, 0x01, 0x01, 0x01};
+
+    ref->seq_num++;
+    target = sf::IpAddress(htonl(Common::BitCast<u32>(ipdata.destination_addr)));
+    ref->tcp_socket.connect(target, ntohs(tcpdata.destination_port));
+    ref->ready = false;
+    ref->ip = Common::BitCast<u32>(ipdata.destination_addr);
+
+    ref->tcp_buffers[0].data = result.Build();
+    ref->tcp_buffers[0].seq_id = ref->seq_num - 1;
+    ref->tcp_buffers[0].tick = GetTickCountStd() - 900;  // delay
+    ref->tcp_buffers[0].used = true;
+  }
+  else
+  {
+    // data packet
+    if (ref == nullptr)
+      return;  // not found
+
+    const int size = ntohs(ipdata.total_len) - ipdata.DefinedSize() - tcpdata.GetHeaderSize();
+    const u32 this_seq = ntohl(tcpdata.sequence_number);
+
+    if (size > 0)
+    {
+      // only if data
+      if ((int)(this_seq - ref->ack_num) >= 0 && data.size() >= size)
+      {
+        ref->tcp_socket.send(data.data(), size);
+        ref->ack_num += size;
+      }
+
+      // send ack
+      WriteToQueue(BuildAckFrame(ref));
+    }
+    // update windows size
+    ref->window_size = ntohs(tcpdata.window_size);
+
+    // clear any ack data
+    if (tcpdata.properties & ntohs(TCP_FLAG_ACK))
+    {
+      const u32 ack_num = ntohl(tcpdata.acknowledgement_number);
+      for (auto& tcp_buf : ref->tcp_buffers)
+      {
+        if (!tcp_buf.used || tcp_buf.seq_id >= ack_num)
+          continue;
+        Common::TCPHeader* tcppart = (Common::TCPHeader*)&tcp_buf.data[0x22];
+        const u32 seq_end =
+            (u32)(tcp_buf.seq_id + tcp_buf.data.size() - ((tcppart->properties & 0xf0) >> 2) - 34);
+        if (seq_end <= ack_num)
+        {
+          tcp_buf.used = false;  // confirmed data received
+          if (!ref->ready && !ref->tcp_buffers[0].used)
+            ref->ready = true;
+          continue;
+        }
+        // partial data, adjust the packet for next ack
+        const u16 ack_size = ack_num - tcp_buf.seq_id;
+        const u16 new_data_size = (u16)(tcp_buf.data.size() - 0x36 - ack_size);
```

Comment:
`0x36` should have a proper name and doesn't take into account the IP/TCP options, if it's the TCP data offset.